From 024fff64113fa5cfae7e505ab3fb0f17765fd2fd Mon Sep 17 00:00:00 2001
From: mlugg <mlugg@mlugg.co.uk>
Date: Sat, 11 Oct 2025 11:10:33 +0100
Subject: [PATCH 004/133] std.posix: panic on unexpected error in `munmap`
X-Developer-Signature: v=1; a=openpgp-sha256; l=1173; i=xtex@astrafall.org;
 h=from:subject; bh=KDPCdDS4LGjFANiEzDlNe+IM2IpT7XOov1a/e10JSs0=;
 b=owGbwMvMwCW2U4Ij7wZL9ETG02pJDJn7cn/826h49UzySq5nrC8X3Jt66eDVs7p3D96o6tFuT
 hSba22l1TGRhUGMi8FSTJGlyLDBm1UnnV90WbkszBxWJpAh0iINDAwMDCwMfLmJeaVGOkZ6ptqG
 eoaGOkY6RgxcnAIw1YlrGRk6FWfxvJn48J1JlSKr0DRtnria3+yCLvxXuV0nnZsom1bFyLBf2vZ
 IpKRB34X9vieiFx2UTjRyP/79Z0iJnctrTZnGOVwA
X-Developer-Key: i=xtex@astrafall.org; a=openpgp;
 fpr=7231804B052C670F15A6771DB918086ED8045B91

This is to help diagnose #25498. We can't use `unexpectedErrno` here,
because `std.posix.munmap` is infallible. So, when the flag is set to
report unexpected errnos, we just call `std.debug.panic` to provide
details instead of doing `unreachable`.

Pushing straight to master after running checks locally; there's no
point waiting for CI on the PR just for this.

(cherry picked from commit 923ddd94a1419abfa9ab95718a7b7d2a8da7a5ec)
---
 lib/std/posix.zig | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/lib/std/posix.zig b/lib/std/posix.zig
index 85b0b331a092..3c7d52f773a5 100644
--- a/lib/std/posix.zig
+++ b/lib/std/posix.zig
@@ -4873,7 +4873,9 @@ pub fn munmap(memory: []align(page_size_min) const u8) void {
         .SUCCESS => return,
         .INVAL => unreachable, // Invalid parameters.
         .NOMEM => unreachable, // Attempted to unmap a region in the middle of an existing mapping.
-        else => unreachable,
+        else => |e| if (unexpected_error_tracing) {
+            std.debug.panic("unexpected errno: {d} ({t})", .{ @intFromEnum(e), e });
+        } else unreachable,
     }
 }
 
-- 
2.52.0

